### PR TITLE
Fix devfile component preferences deserialization through GSON

### PR DESCRIPTION
**Pull request by mkuznyetsov**
### What does this PR do?
Add TypeAdapterFactory for GSON in DtoFactory, that would support Serializable type, that we have in Devfile Preferences

**Special Note regarding parsing of numbers:**
It is known problem that by default all numeric types are parsed as `double` in Gson library, resulting in conversion of integer numbers like `1` into decimal point form `1.0` and rounding dome double type values. [1], [2].  This is a not a problem for JSON itself, since it is doesn't have strict typing, and as a numeric, those two values are identical. But to preserve user input maximally, we're added code that tries to parse value as an `Long` first, and than if wasn't succesfull, fallback to the `Double` form. This preserves the integer values, but may result that decimal zeroes may be trimmed, e.g. `1.00` will be converted into `1`, which is for sure much more rare form of defining integer numbers.

[1] https://github.com/google/gson/issues/1084
[2] https://stackoverflow.com/questions/24926786/make-gson-deserialize-numbers-to-integers-or-doubles
### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/14289

### CODE REVIEW

**Comment by skabashnyuk** [comment]
In general looks ok to me. @mshaposhnik @metlos @sleshchenko wdyt?

**Comment by mshaposhnik** [comment]
Ok if it fixes the problem

**Review by metlos — APPROVED** [review]

**Comment by mkuznyetsov** [comment]
@sleshchenko @mshaposhnik @skabashnyuk @metlos 
so there was a problem with this approach, using default ObjectAdapter all numbers would deserialize to long, even if they could be just integer. Unlike Jackson deserializer, which guesses the correct types in this case.

(there are numbers of issues in GSON repo for this, like this one https://github.com/google/gson/issues/1084)

I'm not sure whether it's critical in our case, but I thought it might be important to have our parsers to behave the same way, so I added a custom type adapter with the last commit to fix this problem in the simplest option.

So the question is, do we need to bother with this at all, and if are there any better alternatives to it

File: core/che-core-api-dto/src/main/java/org/eclipse/che/dto/server/SerializableInterfaceAdapterFactory.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+      objectAdapter.write(out, value);
+    }
+
+    @Override
+    public Object read(JsonReader in) throws IOException {
+      JsonToken token = in.peek();
+      if (token.equals(JsonToken.NUMBER)) {
+        String str = in.nextString();
+        try {
+          return Integer.parseInt(str);
```
Comment (skabashnyuk) [inline core/che-core-api-dto/src/main/java/org/eclipse/che/dto/server/SerializableInterfaceAdapterFactory.java]: what if Number num = NumberFormat.getInstance().parse(myNumber);?

Comment (mkuznyetsov) [reply core/che-core-api-dto/src/main/java/org/eclipse/che/dto/server/SerializableInterfaceAdapterFactory.java]: ok, it works

**Review by skabashnyuk — APPROVED** [review]

**Comment by metlos** [comment]
the latest master needs to be merged into this PR for the happy paths to succeed.

**Comment by mshaposhnik** [comment]
ci-build

**Pull request merged**